components/zcash_address/src/kind/unified.rs

```diff
@@ -69,51 +100,214 @@ impl Typecode {
     type Error = TryFromIntError;
+
     fn try_from(t: Typecode) -> Result<Self, Self::Error> {
         u32::from(t).try_into()
     }
 }
 
-impl Typecode {
-    fn is_transparent(&self) -> bool {
-        // Unknown typecodes are treated as not transparent for the purpose of disallowing
-        // only-transparent UAs, which can be represented with existing address encodings.
-        matches!(self, Typecode::P2pkh | Typecode::P2sh)
+/// An enumeration of known Unified Metadata Item types.
+///
+/// Unknown MUST-understand metadata items are NOT represented using this type, as the presence of
+/// an unrecognized metadata item with a typecode in the `MUST-understand` range will result in a
+/// parse failure, instead of the construction of a metadata item.
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub enum MetadataItem {
+    /// The expiry height for a Unified Address or Unified Viewing Key
+    ExpiryHeight(u32),
+    /// The expiry time for a Unified Address or Unified Viewing Key
+    ExpiryTime(u64),
+    /// A Metadata Item with an unrecognized Typecode. MUST-understand metadata items are NOT
+    /// represented using this type, as the presence of an unrecognized metadata item with a
+    /// typecode in the `MUST-understand` range will result in a parse failure.
+    Unknown { typecode: u32, data: Vec<u8> },
+}
+
+impl MetadataItem {
+    /// Parse a metadata item for the specified metadata typecode from the provided bytes.
+    pub fn parse(
+        revision: Revision,
+        typecode: MetadataTypecode,
+        data: &[u8],
+    ) -> Result<Self, ParseError> {
+        match (revision, typecode) {
+            (Revision::R1, MetadataTypecode::ExpiryHeight) => data
+                .try_into()
+                .map(u32::from_le_bytes)
+                .map(MetadataItem::ExpiryHeight)
+                .map_err(|_| {
+                    ParseError::InvalidEncoding(
+                        "Expiry height must be a 32-bit little-endian value.".to_string(),
+                    )
+                }),
+            (Revision::R1, MetadataTypecode::ExpiryTime) => data
+                .try_into()
+                .map(u64::from_le_bytes)
+                .map(MetadataItem::ExpiryTime)
+                .map_err(|_| {
+                    ParseError::InvalidEncoding(
+                        "Expiry time must be a 64-bit little-endian value.".to_string(),
+                    )
+                }),
+            (Revision::R1, MetadataTypecode::Unknown(typecode)) => Ok(MetadataItem::Unknown {
+                typecode,
+                data: data.to_vec(),
+            }),
+            (Revision::R1, MetadataTypecode::MustUnderstand(tc)) => {
+                Err(ParseError::NotUnderstood(tc))
+            }
+            (Revision::R0, _) => {
+                // The entire R0 metadata range has now been reserved for MUST-understand metadata
+                // in R1, so no R0 metadata range now exists.
```

Comment:
I disagree: on [line 126](https://github.com/zcash/librustzcash/pull/1135/files#diff-55c7c102f533ffc792f55a3517e036e5e41f08c1ee669b688a91d9ac90162b6fR126) we parse 0xC0..=0xDF as a metadata typecode, independent of revision. This range should be allowed without error, because either possible interpretation of the spec (that these are unknown data typecodes, or that they are retrospectively changed to unknown and *not* MUST-understand metadata typecodes) would lead to allowing them without error.